rollouts/controllers/remoterootsync_controller.go

```diff
@@ -145,42 +156,66 @@ func (r *RemoteRootSyncReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
-	clusterRef := &remoterootsync.Spec.ClusterRef
-	dynCl, err := r.getDynamicClientForCluster(ctx, clusterRef)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+	syncStatus, syncError := r.syncExternalSync(ctx, &remoterootsync)
 
-	if err := r.patchRootSync(ctx, dynCl, req.Name, &remoterootsync); err != nil {
+	if err := r.updateStatus(ctx, &remoterootsync, syncStatus, syncError); err != nil {
+		logger.Error(err, "Failed to update status")
 		return ctrl.Result{}, err
 	}
 
-	r.setupWatches(ctx, remoterootsync.Name, remoterootsync.Namespace, remoterootsync.Spec.ClusterRef)
+	return ctrl.Result{}, syncError
+}
+
+func (r *RemoteRootSyncReconciler) syncExternalSync(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync) (string, error) {
+	syncName := rrs.Name
+	clusterRef := &rrs.Spec.ClusterRef
 
-	syncStatus, err := checkSyncStatus(ctx, dynCl, req.Name)
+	dynCl, err := r.getDynamicClientForCluster(ctx, clusterRef)
 	if err != nil {
-		return ctrl.Result{}, err
+		return "", fmt.Errorf("failed to create client: %w", err)
 	}
 
-	if err := r.updateStatus(ctx, &remoterootsync, syncStatus); err != nil {
-		logger.Error(err, "Failed to update status")
-		return ctrl.Result{}, err
+	if err := r.patchRootSync(ctx, dynCl, syncName, rrs); err != nil {
+		return "", fmt.Errorf("failed to create/update sync: %w", err)
 	}
 
-	return ctrl.Result{}, nil
+	r.setupWatches(ctx, rrs.Name, rrs.Namespace, rrs.Spec.ClusterRef)
+
+	syncStatus, err := checkSyncStatus(ctx, dynCl, syncName)
+	if err != nil {
+		return "", fmt.Errorf("faild to check status: %w", err)
+	}
+
+	return syncStatus, nil
 }
 
-func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, syncStatus string) error {
+func (r *RemoteRootSyncReconciler) updateStatus(ctx context.Context, rrs *gitopsv1alpha1.RemoteRootSync, syncStatus string, syncError error) error {
 	logger := klog.FromContext(ctx)
 
-	// Don't update if there are no changes.
-
 	rrsPrior := rrs.DeepCopy()
+	conditions := &rrs.Status.Conditions
 
-	rrs.Status.SyncStatus = syncStatus
-	rrs.Status.ObservedGeneration = rrs.Generation
+	if syncError == nil {
+		rrs.Status.SyncStatus = syncStatus
+		rrs.Status.SyncCreated = true
+
+		meta.RemoveStatusCondition(conditions, conditionReconciling)
```

Comment:
Can you please add a TODO here:
`Reconciling` or `Stalled` condition should be updated (or externalSync conditions should be added) by examining `syncStatus`. `SyncStatus` could have values `Stalled, Reconciling, Error, Pending` even when `syncError` is nil.

Reply:
I don't think we want actually do this with these being [kstatus conditions](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus). From my understanding, the `Reconciling` and `Stalled` conditions (in the case of the remote root sync) should only consider if the root sync is created and up to date with the latest spec as defined by the remote root sync, regardless of the sync status of the root sync. When the `Reconciling` condition is set to true it implies that the Remote Root Sync still needs to reconcile the root sync with the lates spec from the remote root sync. @mortent can you confirm?

Reply:
`if the root sync is created and up to date with the latest spec as defined by the remote root sync`

You are right and that's where my comment come from. If you look at `checkSyncStatus`, it can return `Pending` status without any error when `generation != observedGeneration` (meaning rootsync is *not* up to date with the latest spec of rootsync). I think for other values `Stalled, Reconciling, Error`, we will have to create an extra condition to track `externalSyncStatus` or something.

Happy to chat offline :)